Assets/MRTK/SDK/Experimental/SpatialAwareness/SurfaceMeshesToPlanes.cs

```diff
@@ -0,0 +1,473 @@
+#if PLANE_FINDING_PRESENT
+        private float snapToGravityThreshold = 5.0f;
+#endif // PLANE_FINDING_PRESENT
+
+#endregion
+
+#region Public Methods
+
+        /// <summary>
+        /// Gets all active planes of the specified type(s).
+        /// </summary>
+        /// <param name="planeTypes">A flag which includes all plane type(s) that should be returned.</param>
+        /// <returns>A collection of planes that match the expected type(s).</returns>
+        public List<GameObject> GetActivePlanes(SpatialAwarenessSurfaceTypes planeTypes)
+        {
+            List<GameObject> typePlanes = new List<GameObject>();
+
+            foreach (PlaneWithType planeWithType in activePlanes)
+            {
+                if ((planeTypes & planeWithType.Type) == planeWithType.Type)
+                {
+                    typePlanes.Add(planeWithType.Plane);
+                }
+            }
+
+            return typePlanes;
+        }
+
+        /// <summary>
+        /// Runs background task to create new planes based on data from mesh observers
+        /// </summary>
+        public void MakePlanes()
+        {
+            if (!makingPlanes)
+            {
+                makingPlanes = true;
+                tokenSource = new CancellationTokenSource();
+                _ = MakePlanes(tokenSource.Token).ConfigureAwait(true);
+            }
+        }
+
+#endregion
+
+#region Private Methods
+
+        private void Start()
+        {
+            activePlanes = new List<PlaneWithType>();
+
+            if (planesParent == null)
+            {
+                planesParent = new GameObject("SurfaceMeshesToPlanes");
+            }
+            planesParent.transform.position = Vector3.zero;
+            planesParent.transform.rotation = Quaternion.identity;
+
+            MakePlanes();
+        }
+
+        private void OnDestroy()
+        {
+            Destroy(PlanesParent);
+            tokenSource?.Cancel();
+        }
+
+        /// <summary>
+        /// Task to analyze surface meshes to find planes and create new 3D cubes to represent each plane.
+        /// </summary>
+        /// <param name="cancellationToken">Token that allows cancellation of an asynchronous process.</param>
+        /// <returns>Yield result.</returns>
+        private async Task MakePlanes(CancellationToken cancellationToken)
+        {
+            await new WaitForUpdate();
+
+#if PLANE_FINDING_PRESENT
+
+            List<PlaneFinding.MeshData> meshData = new List<PlaneFinding.MeshData>();
+            List<MeshFilter> filters = new List<MeshFilter>();
+
+            var spatialAwarenessSystem = CoreServices.SpatialAwarenessSystem;
+            if (spatialAwarenessSystem != null)
+            {
+                GameObject parentObject = spatialAwarenessSystem.SpatialAwarenessObjectParent;
+
+                // Get mesh filters from SpatialAwareness Mesh Observer
+                foreach (MeshFilter filter in parentObject.GetComponentsInChildren<MeshFilter>())
```

Comment:
How does this script aquire these meshes?

Reply:
Should we make it a mesh event listener?

Reply:
I am going to leave that up to the developer. In Holograms 230, we collected meshes for n seconds and then called the MakePlanes method. For a demo, I am considering adding a voice command to generate meshes.

Reply:
Ahh I see. I missed the part where this script was specifically grabbing the `SpatialAwarenessObjectParent`.

Reply:
>I am considering adding a voice command to generate meshes.

Yeah, a voice command or even just a button seems reasonable